File: 572. Subtree of Another Tree/sol.py


class TreeNode:
    def __init__(self, val: int = 0, left: 'Optional[TreeNode]' = None, right: 'Optional[TreeNode]' = None):
        self.val = val 
        self.left = left 
        self.right = right 

class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        def isSameTree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
            stack = [[p, q]]
            while len(stack) > 0:
                p, q = stack.pop()
                if p or q:
                    if not p or not q:
                        return False
                    if p.val != q.val:
                        return False
                    stack.append([p.left, q.left])
                    stack.append([p.right, q.right])
            return True


        stack = [root]
        node = root
        while len(stack) > 0 :
            node = stack.pop()
            if node:
                if node.val == subRoot.val:
                    if isSameTree(node, subRoot):
                        return True
                stack.append(node.left)
                stack.append(node.right)
        
        return False

File: 572. Subtree of Another Tree/sol.go
package subtreeofanothertree 

type TreeNode struct {
	Val int 
	Left *TreeNode 
	Right *TreeNode 
}

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	stack := []*TreeNode {root}
    var node *TreeNode
    for len(stack) > 0 {
        index := len(stack) - 1
        node = stack[index]
        stack = stack[:index]
        if node != nil {
            if node.Val == subRoot.Val {
                if isSameTree(node, subRoot) { return true}
            }
            stack = append(stack, node.Left, node.Right)
        }
    }
    return false
}


func isSameTree(p *TreeNode, q *TreeNode) bool {
    stack := [][]*TreeNode { []*TreeNode {p, q}}
    for len(stack) > 0 {
        index := len(stack) - 1
        a, b := stack[index][0], stack[index][1]
        stack = stack[:index]
        if a != nil || b != nil {
            if a == nil || b == nil {
                return false
            }
            if a.Val != b.Val {
                return false
            }
            stack = append(stack, []*TreeNode {a.Left, b.Left}, []*TreeNode {a.Right, b.Right})
        }
    }
    return true
}